File: 54-spiral-matrix/54-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        direction = [(-1,0),(0,-1),(1,0),(0,1)]
        m, n = len(matrix), len(matrix[0])
        point = (0,0)
        res = []
#        visited = [[False]*n for i in range(m)]
        for i in range(m*n):
            row, col = point[0], point[1]
            res.append(matrix[row][col])
#            visited[row][col] = True
            matrix[row][col] = -101

            nxt_row, nxt_col = row+direction[-1][0], col + direction[-1][1]
            if  0<= nxt_row < m and 0<= nxt_col < n and matrix[nxt_row][nxt_col] != -101:
                point =(row+direction[-1][0],col + direction[-1][1] )
            else:
                direction = [direction.pop()] + direction
                point =(row+direction[-1][0],col + direction[-1][1] )
        return res
                
                
            
            

File: 54-spiral-matrix/54-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        LinkedList<int[]> directions = new LinkedList<>(Arrays.asList(new int[][]{{0,1},{1,0},{0,-1},{-1,0}}));
        List<Integer> res = new LinkedList<>();
        int nRow = matrix.length, nCol = matrix[0].length;
        int x = 0, y = 0;
        res.add(matrix[0][0]);
        matrix[0][0] = -101;
        for(int i = 0; i < nRow*nCol - 1; i++){
            while(true){
                int[] curDir = directions.peekFirst();
                int newx = x + curDir[0], newy = y + curDir[1];
                if(newx < 0 || newx >= nRow || newy < 0 || newy >= nCol || matrix[newx][newy] == -101){
                    directions.pollFirst();
                    directions.addLast(curDir);
                }else{
                    
                    res.add(matrix[newx][newy]);
                    x = newx;
                    y = newy;
                    matrix[newx][newy] = -101;
                    break;
                }
            }   
        }
        return res;
        
        
    }
}